Rework solution configuration and ST CMSIS/HAL integration

josesimoes opened the pull request:
This PR addresses the discussions in #405 and also #429.
A couple of warnings:
- it's a work in progress
- I'm aware that it's huge but couldn't find a better way to deal with all the changes because it touches a lot of stuff
- I know it's not perfect, but it's a start 

Having said that: it's a proposal open for review, discussion and criticism.

What's included (and in the works)
1. Add PowerShell script to download and "install" ST CMSIS and HAL files (this addresses the question of not having third party code in the repository).
2. The script is called when building a solution to make sure everything is available (this addresses the question of dealing with automated build processes).
3. Reworked the way the hardware features are added to a solution to make developer's life easier and not requiring a deep knowledge of what is "under the hood".
   1. The device series/variant is moved from platform selector to solutions settings.
   2. TinyBooter and TinyCLR each have a settings file to allow granular selection of what features are to be added on each flavor. These settings files merge the solution settings and each is the one that goes into the respective proj file (instead of the global one as before).
   3. These settings files have a simple True/False selector to choose what should be added. These properties latter reflect what is compiled/linked in the final image. Having this selection here and as a property allows them to be visible and accessible in the projs, build system and code.
   4. The features property allow that (using #defines) the calls to the HAL init/deinit functions are added or not according to the feature selection. 
   5. There is a readme file with the instructions on how to build a new solution.
   6. The  TinyBooter and TinyCLR projs are organized in a way that they don't need to be (almost...) touched when building a new solution. Unless a very particular change is required, of course.
   7. There is a HAL project (one for CLR and another for loaders) that includes all the features.
   8. All the stubs are now implemented as 'weak' functions. This makes things easier because it allows that all the stubs can now be compiled and always added by default because the linker will replace  the 'weak' ones with the 'strong' implementation (when it exists) and the optimizations will remove the 'dead' code when they are not used.
   9. All this makes it very easy to add new series and MCUs (at least for ST parts, right now there is only the HAL for the F4xx series but any other series can be added with minimal effort by using the ST Cube files and taking advantage on the new organization).
4. The "new" HAL is total based in the CMSIS and HAL code supplied by ST.
5. The individual projects for the HAL features are now based in the ST HAL and follow the ST guidelines and philosophy.
6. The hardware configuration is now simpler because it uses the HAL naming and defines. This makes it easier to add new series/variants becuase the code knows what is available and what isn't and it adapts accordingly. E.g how many and which USARTS, SPI port, which pins belong where, etc.
   (GPIO, USART, SPI, I2C, power, flash and time are implemented; DA, ADC, PWM and USB are on the works)
7. CRC calculations are selectable between the software implementation and hardware based.   
8. Most of the configurations and HAL options are now in the system_stm32f4xx.c in the solution folder, replacing what was in platform selector.
9. USB support (using ST middleware libs) should be much easier for both device and host. We are close to have USB connection for MFDeploy. It currently enumerates the NETMF device and receives the debug packets. The last part (replying to MFDeploy) is in the works.  
10. The scatterfiles were reworked to use of the newlib nano and implement the stack/heap. The managed heap size is now automatic.
11. The "simple heap" was removed because the modern compilers can provide that and we can free NETMF from doing it. (There is still a small remapping required for the crypto libs).

Final notes:
- The changes are being included in both MCBSTM32F400 and STM32F4DISCOVERY. We only have available the STM32F4DISCOVERY board so we can't test the other hardware.
- Right now only the TinyBooter is compiling and tested (except USB).

josesimoes committed:
Update CMSIS to 4.5

- checked implementation of osXxxxDef objects
- need to add cpp define for __FPU_PRESENT because CMSIS code has
changed and the FP checks on previous versions that produced warning are
now errors

josesimoes committed:
revert gitignore to match upstream dev

josesimoes committed:
remove licenses and other txt files from crypto libs

josesimoes committed:
update with crypto lib licenses and other txt

josesimoes committed:
Merge branch 'fix-for-signing-empty-files' into dev

josesimoes committed:
Add extension to image files generated with GCC.

(this makes them ready to be uploaded with ST-Link app, DFU manager or similar)

josesimoes committed:
Merge branch 'dev' of https://github.com/NETMF/netmf-interpreter into dev

josesimoes committed:
Revert changes in adding extension to images built with GCC

josesimoes committed:
Tweak gcc targets for GCC 5-2015-q4-major

- replace gcc with g++ fro CPP files
- add extra information to debug flavors and axfdump file
- add debug flavored configurations for arch (linked file)

josesimoes committed:
Revert "Tweak gcc targets for GCC 5-2015-q4-major"

This reverts commit feaf5ad8eae4274a4ed2b097a531128a6a5c6314.

josesimoes committed:
Merge branch 'dev' of https://github.com/NETMF/netmf-interpreter into dev

josesimoes committed:
Merge branch 'dev' of https://github.com/NETMF/netmf-interpreter into dev

josesimoes committed:
Add folder to work as placeholder for STM32Cube packages

- add PowerShell script to download and install SMT32F4 support pack from ST web site
(work in progress as the destination folders and final layout are still unknow)

josesimoes committed:
Remove comment so script actually downloads the pack file

josesimoes committed:
Update PowerShell script

- script is now generic to be able to handle all available STM32 series
- add code to copy source files to correct destination directories
- rename script

josesimoes committed:
Improve script

- add/reword comments
- move download code into if()

josesimoes committed:
Several improvement in script

- now accepts series names and target version as parameters
- add validation for parameters
- add code to check if script is running from repository home or being called from a build project and adjusting working path accordingly
- add standard help to script

josesimoes committed:
Merge branch 'dev' of https://github.com/NETMF/netmf-interpreter into CMSIS_STM32F4

josesimoes committed:
Improvements in script

- correct pack target version validation
- correct documentation
- remove debug code

josesimoes committed:
Improvements in script

- replace several calls with PowerShell comandlets to improve eficiency
- source folder is now deleted after copy
- destination directories are now cleared before copying files

josesimoes committed:
Update feedback messages

josesimoes committed:
Rename SD card commands because some colide with STM32 HAL declarations

josesimoes committed:
Correct bug in script

josesimoes committed:
Rename FLASH in HAL_SYSTEM_CONFIG because it was coliding with a CMSIS declaration with the same name

josesimoes committed:
Rename variable because it was coliding with a CMSIS declaration with the same name

sjmneves committed:
Added project for HAL CMSIS GPIO

josesimoes committed:
Add CMSIS includes for STM32F4 series

josesimoes committed:
Remove custom heap (as discussed in issue #417)

- remove references to SimpleHeap projects
- remove calls to SimpleHeap config and init
- remove Simple Heap declarations from TinyHal.h
- replace private_malloc and private_free with standard C malloc and free
- implement redirect functions for private_malloc and private_free because they were being referenced from the pre-compiled crypto libs
- rework scatter files to have a managed heap and standard heap that are managed from C standard code

sjmneves committed:
Merge branch 'CMSIS_STM32F4' of https://github.com/Eclo/netmf-interpreter into CMSIS_STM32F4

sjmneves committed:
Changed ifdef block for STM32 families

- now looking for bit definition and not for the family

sjmneves committed:
Added project for HAL CMSIS I2C

sjmneves committed:
Added project for HAL CMSIS SPI

sjmneves committed:
Rename file i2c_functions.cpp to I2C_functions.cpp

josesimoes committed:
Correct comments and projects cleanup

josesimoes committed:
Add CMSIS HAL project for STM32F4xx series

- This projects glues the HAL and PAL for pure CMSIS
- Includes new approach for features selection at solution level
- Minor changes in several files to make them suitable for CMSIS

josesimoes committed:
Implemented new "feature selection" approach at solution level

- change current 'solution'.settings to accomodate new approach
- change current 'solution'.settings and others to use new CMSIS HAL
- change scatter files to remove simple heap, add auto size for managed heap, add C/C++ standard heap
- change TinyCLR and TinyBooter projects to use new feature selection approach
- add call to PowerShell script to install CMSIS package before building solution

*** WORK IN PROGRESS ***
- tested with DISCOVERY4 TinyBooter and GCC

josesimoes committed:
Update script to copy USB device library files

sjmneves committed:
Change define for SPI ports

now user must define available SPI ports in a preloaded define in platform_selector.h

josesimoes committed:
Improve script

- renames library souce files to .cpp so they get compiled as CPP code (required in order to include tinyhal.h
- delete template files from library source code

josesimoes committed:
Improvements in CMSIS HAL

- moved system_stm32f4xx.c to HAL folder
- add feature setting for cpu cache, rework solution projects and add if/defs on the code
- start working on CMSIS time
- add missing if/defs in code for LCD feature
- improvements in system clock config

josesimoes committed:
Correct indentations

sjmneves committed:
Added project for HAL CMSIS USART

sjmneves committed:
Missing file from last commit, defines needed

josesimoes committed:
Add back wrongly removed feature project for SPI and I2C

josesimoes committed:
Implement CMSIS time to provide NETMF time base

josesimoes committed:
Rework HAL

- add file with interrupt handlers
- removed NVIC call to set priority of SysTickIRS (HAL_Init is already taking care of this)
- removed code to config HAL_SYSTICK in System_Clock_Config (HAL_Init is already taking care of this)
- remove definition of several interrupt handlers from the startup code (GCC is not replacing the weak refs with the hard ones, maybe a bug?)
- moved asm files to the bottom
- correct indentation

josesimoes committed:
Correct comment

josesimoes committed:
Enable MCU debug when not building RTM

sjmneves committed:
- Handle USART/UART interupts

josesimoes committed:
Implement CMCSIS power

- this includes a different (simplified) implemention to be used in TinyBooter
- manages CPU power modes changes and reset
- remove repeated code in tinybooter.cpp
- remove unused stubs from TinyBooter projects

josesimoes committed:
Cleanup comments

josesimoes committed:
Cleanup comments

josesimoes committed:
Remove Groups from Time proj

josesimoes committed:
Add function to get USART interrupt enable state (required for PAL)

josesimoes committed:
Add project for USB device (WinUSB)

- this is a work in progress
- right now it's able to enumerate the device as WinUSB device, set the NETMF class for debug and reply with the appropriate descriptors to be found by MFDeploy

josesimoes committed:
Correct comment

josesimoes committed:
Project for CMSIS USB implementation

- this is a work in progress

josesimoes committed:
Add readme with steps on how to configure a solution

josesimoes committed:
Remove unused file

josesimoes committed:
Correct file name case

josesimoes committed:
Implement CMSIS Flash driver

- update solution projects and block storage definitions accordingly

josesimoes committed:
Correct implementation of CMSIS for MCBSTM32F400 solution

- add missing hal_conf file
- add missing groups in global dotnetmf proj

sjmneves committed:
Make all function "weak"

sjmneves committed:
Turn all functions "weak"

sjmneves committed:
Turn all function "weak"

sjmneves committed:
Turn all function "weak"

sjmneves committed:
Turn all function "weak"

sjmneves committed:
Turn all function "weak"

sjmneves committed:
Turn all function "weak"

sjmneves committed:
Turn all function "weak"

josesimoes committed:
Correct some typos and formating

josesimoes committed:
Remove static from methods declaration

(weak declarations must be public, otherwise the compiler will complain about that)

josesimoes committed:
Add weak attribute to methods in stub projects

josesimoes committed:
Correct issues with build

josesimoes committed:
Add new feature selection: watchdog

- add feature to features settings files
- update solution projects accordingly
- rework HAL projects to always add all stubs projects (now have weak references and will be replaced by actual implementation when appropriate)
- rework solution projects to include mandatory and stubs projects

josesimoes committed:
Add new feature selection: analog (ADC)

(CMSIS not available yet)

josesimoes committed:
Remove CPU EBIU and interrupt management from MCBSTM32F400 projects

josesimoes committed:
Add new feature selection: PWM

(CMSIS not available yet)

josesimoes committed:
Add new feature selection: DA

(CMSIS not available yet)

josesimoes committed:
Remove unused code

josesimoes committed:
Move Itm0GenericPort and g_GenericPorts declaration to start code in each solution

- this was the last usable code in solution\DeviceCode\Init\ proj
- remove unused code

josesimoes committed:
Add new feature selection: sockets

- add compiler defs to remove calls to socket methods in ComDirector when socket feature is not available
(CMSIS not available yet)

josesimoes committed:
Add compiler defs to remove calls to USART methods in ComDirector when USART feature is not available

josesimoes committed:
Add compiler defs to remove calls to Watchdog methods when Watchdog feature is not available

josesimoes committed:
Add missing lib reference for I2C in bootloader projects

josesimoes committed:
Add new feature selection: time service

josesimoes committed:
Reorganized HAL/PAL projects inclusion order and arrangement for clarity

josesimoes committed:
Further reorganization of PAL/HAL inclusion in solution projects

josesimoes committed:
Add new feature selection: PAL event

josesimoes committed:
Add project with HAL interface for Cortex-M CRC calculation unit

josesimoes committed:
Add comments

josesimoes committed:
Add new features selection: Software CRC and Cortex-M CRC

sjmneves committed:
Rework GPIO to make it CMSIS compliant

sjmneves committed:
Clean code

sjmneves committed:
Rework Initialize and Uninitialize functions to be compliant with CMSIS

sjmneves committed:
Added arrays size validation

sjmneves committed:
Rework code to make it CMSIS compliant

josesimoes committed:
Set NVIC priorities for USART modules

josesimoes committed:
Correct implementation of interface to hardware CRC calculation unit

- fully working for 0 init value and when size is multiple of 4
- still needs work for the remanining situations

sjmneves committed:
Correct include path for Analog, PWM and DA features

sjmneves committed:
Merge branch 'CMSIS_STM32F4' of https://github.com/Eclo/netmf-interpreter into CMSIS_STM32F4

sjmneves committed:
Added CMSIS Analog, PWM, DA drivers (work in progress)

- updated I2C code (work in progress)
- added new defines to platform_selector (work in progress)

josesimoes committed:
Add NVIC priorities to processor selector

josesimoes committed:
Update script to copy USB midleware to correct target folder

josesimoes committed:
Add ignore path to USB midleware library folders

josesimoes committed:
Update TICK_INT_PRIORITY to make it usable in other preprocessor defines

josesimoes committed:
Move STM32F4xx HAL to correct folder

- update paths
- delete unused folders

josesimoes committed:
Merge branch 'dev' of https://github.com/NETMF/netmf-interpreter into CMSIS_STM32F4

josesimoes committed:
Cleanup unused defines

sjmneves committed:
Complete CMSIS DA driver

sjmneves committed:
Merge branch 'CMSIS_STM32F4' of https://github.com/Eclo/netmf-interpreter into CMSIS_STM32F4

sjmneves committed:
Complete CMSIS Analog driver

sjmneves committed:
Added missing lib I2C_pal

smaillet-ms commented:
You'll need to split this up into smaller PRs before we can consider it, there are 113 commits across 200+ files and it brings the GitHub site to it's knees when attempting to review or provide feedback.

smaillet-ms commented:
Also FYI, there are already powershell scripts for downloading the CMSIS and binary tools. And I'm working on another to download and install the crypto libs (Though that can't actually be done for an automated build service running in a restricted account it will allow building a single powershell script to get all the pre-requisites for individual developers)

josesimoes commented:
@smaillet-ms  I can go and try to squash that into a single commit... does it help? 
Because this is a really big change touching a lot of files. Not sure how small/atomic I can make those smaller PRs for each individual one to make sense...

Don't know if you are familiar with it but there is an awesome tool for Git: [Git Extensions](https://github.com/gitextensions/gitextensions). It works great for this kind of work that tends to be impossible to manage on the web interface. You may want to give it a try. There is a tool for working with PRs (no matter how big they are!) :wink: 

Regarding the PowerShell script: I'm aware of it The thing is that that script downloads the CMSIS "general" distribution. The one that is there on the PR is different: it downloads the CMSIS implementation for STM32 devices and the ST HAL files. 

As for running the ps scripts in a build service I believe it can be accomplished by signing the scripts. I could have signed the one in the PR but it would have our company signature. I believe it should be possible to include the ps scripts in the NETMF sign process and have them signed with Microsoft certificates.

sjmneves committed:
Complete CMSIS I2C driver

sjmneves committed:
Correct path

smaillet-ms commented:
PS scripts in the build service aren't the issue - that's actually built into the VSO builds we are now using, it's the fact that we are required to distribute the crypto libs as an MSI that's the problem. When a build service is set up it doesn't (and frankly shouldn't be) run with permissions to install software on the server, so it can't run an MSI. 

So you are also pulling the ST packs etc... Can we hold on this then?

We already have a lengthy discussion on the build system and how we want to solve the problem of these types of dependencies. I've published the start of the [CMSIS-Pack .NET library](https://github.com/NETMF/CMSIS.Pack) for managing a pack repository from C# so we can have a new app that is a cross between the pack manager in uVision and the old PKStudio and build that into VS. I'd rather not do that kind of multiple times.

josesimoes commented:
1) This is maybe sound like a dumb question, but... can't you just distribute the crypto libs with the plain files? Without the MSI and requiring an installation? Or even better, as a Nuget package? For me this would address this automated build/install/etc hurdle.

2) Yes that PS script is able to download and install any ST pack for any of the available series.

3) I saw your new CMSIS package manager last Friday but hadn't a chance to look into it yet. From your description above it will be usable from VS at some point, correct? For that to be usable wouldn't it require a way to manage the building a .NETMF Solution as a VS solution?

josesimoes commented:
@smaillet-ms I would be very interested on your comments on the way we are proposing to manage the features availability in a Solution. I really believe it eases (a lot!) regarding what a developer needs to touch and understand about Solutions architecture and not having to touch the prj files and all that.

smaillet-ms commented:
I'll be looking into that on this PR. Please note I'm not in any way trying to suggest there's nothing of value here. There's a lot of different angles and aspects to be considered for a robust architecture neutral system. We've got a system now that was the result of multiple whiplash speed design/implementations and I'd like to avoid repeating any mistakes. Furthermore we've got some bugs that are blocking real world use in SSL and some VB issues to knock out (likely requiring a 4.4.1 release at least). I'd like to hit the bigger build stuff right after those are done.

josesimoes commented:
Understood. 
Know that we are doing this for real world implementations too. Our company develops commercial products based in .NETMF.
We are here because we wanted to contribute back to the community for all the great stuff that we've been receiving so far. And because we love .NETMF and believe it's a truly remarkable platform.
All this started because we needed networked SSL connectivity and we are using GCC.
First approach was to use FreeRTOS instead of RTX, which we did.
Next we've stumbled on the need to update the HAL to the new ST stuff and then the well known issues with networked connectivity and SSL.
In this PR we've gone through most of the HAL stuff. Right now the USB connection to MFDeploy needs a final debug and it will be usable. Next week we'll start adding the network stuff including the LwIP and SSL.
All this to say that we are activelly working on this and we are under a tight schedule for a project we need to deliver. If we are able to mix and match both our teams efforts: excelent. If not, so be it but we really need to keep moving forward with this. At a latter time some of our work maybe usefull.

smaillet-ms commented:
Understood. I hate duplication and re-working something multiple times but sometimes the real world gives you little choice. Having something with some field time on it isn't a bad thing either. It will give you a chance to evaluate the choices you made and see how well they hold up in the real world rather than just on paper. That will help drive to a better overall long term experience.

It should be a lot easier to put NETMF onto existing Cortex-M devices given how much is available in CMSIS now. And for any other processor type there are similar types of libraries available now to work from as well.

josesimoes committed:
Replace Simple heap calls with standard C malloc() and free()

josesimoes committed:
Correct flash field names in HalSystemConfig struct

josesimoes committed:
Add extern declaration for SystemClock_Config

josesimoes committed:
Remove reference to HAL_DECLARE_CUSTOM_HEAP

josesimoes committed:
Remove implementation of HAL_CPU_Sleep (is now in Power project)

josesimoes committed:
Add missing projects and correct identation

josesimoes committed:
Add missing ref to CRC project

josesimoes committed:
Several corrections to TinyCLR

- correct path for settings file
- enabled newlib nano compilation
- add missing projects
- remove unused features for testing
- add missing I2C projects/lib refs

josesimoes committed:
Corrections of I2C projects/lib refs

josesimoes committed:
Revert changes with SimpleHeap

- declaration must be in tinyhal.h oterwise the framework doesn't compile
- remove weak declarations from grahpics and time stubs oterwise the framework doesn't compile

josesimoes committed:
Remove weak declarations from simple heap stubs otherwise the framework doesn't compile

josesimoes committed:
Remove unused code and headers from HAL prjs and HAL configuration file at solution level

sjmneves committed:
Rework GPIO to improve performance and to use only HAL functions

sjmneves committed:
Added define for GPIO port pin structer

josesimoes committed:
Correction and improvements in GPIO HAL

- correct bug with functions being called with GPIO defs instead of pin index
- replace code with ST HAL calls
- correct bug in CPU_GPIO_EnableOutputPin was setting the output state before configuring the pin
- improve GPIO and LEDs definition in platform_selector.h

josesimoes committed:
Correct comment

josesimoes committed:
Correct order of calls to timers init

josesimoes committed:
Further work of USB debug project

- WinUSB classes are working
- USB debug and communication with MFDeploy completed
- Move USB config from solution\device code to solution settings file
- Add line to targets in order to set defines required for USB config compilation
- Rework TinyBooter and TinyCLR proj to include USB debug project/features
- Add FEATURE_USB_DEBUG to solutions features settings files and STM32F4 HAL projects

josesimoes committed:
Add flag to force c11 and c++11 code and missing air quotes in include paths

sjmneves committed:
Correct interrupt handler function names to match declaration of startup code

sjmneves committed:
Add missing extern "C" decoration to interrupt handlers

sjmneves committed:
Move declaration of weak interrupt handlers from startup code to Int_Handlers

josesimoes committed:
Correct property definition to properly add USB config defines

josesimoes committed:
Move startup code to solution folder

(only GCC, still missing RVD version)

josesimoes committed:
Move compiler defines of HAL/PAL features to Microsoft.SPOT.Features.Settings

josesimoes committed:
Rename CRC feature definition to make it clearer

josesimoes committed:
Remove comments from interrupt handlers at startup code

 (weak declaration is now being correctly replaced with strong implementation by linker)

josesimoes committed:
Add missing ifdef FEATURE_LCD wrapping call to lcd_printf

josesimoes committed:
Correct code with wrong call to USB_OpenStream

- see PR #439

josesimoes committed:
Correct comment

josesimoes committed:
Merge branch 'dev' of https://github.com/NETMF/netmf-interpreter into CMSIS_STM32F4

josesimoes committed:
Merge branch 'dev' of https://github.com/NETMF/netmf-interpreter

josesimoes committed:
Missed last commit

josesimoes committed:
Added STM32F429IDISCOVERY solution  from https://github.com/NETMF/netmf-interpreter

josesimoes committed:
Merge branch 'dev' of https://github.com/NETMF/netmf-interpreter into CMSIS_STM32F4

josesimoes committed:
Several updates and improvements in script

- add code and validation to accept all possible versions
- correct download URL
- add code to handle patch packages

josesimoes committed:
Add gitignore for Visual Studio Code

josesimoes committed:
Update CMSIS package number to 1.13.0

josesimoes committed:
Add nosys.specs for new lib nano

josesimoes committed:
Rework code to support debug scenarious

- add support for SWO
- rework code to properly handle pin configs

josesimoes committed:
﻿Correct implementation of cascading timers for NETMF time base

- set proper reload values for timers
- remove wrong/duplicated config code
- correct code to read current ticks from cascaded timers
- correct typos

josesimoes committed:
Update CRC code to match PR #526
